text/0000-builders.md

```diff
@@ -0,0 +1,101 @@
+1. Static files
+
+Deploying this is relatively simple, but certainly not optimised for certain deployment strategies such as serverless (lambda), and often relies on third-party, platform-specific builders.
+
+If Sapper were to produce its output in a more modular fashion, it would facilitate the creation of a number of officially supported "builders" for major platforms, as well as opening the doors for third parties to easily produce builders which targetted other platforms.
+
+Additionally, it would ease the creation of some desirable features:
+
+1. Per-route bundle splitting
+1. [Differential rendering on a per-route basis](https://github.com/sveltejs/sapper/issues/1324) - Allowing SSR, SSG, SPA, [JAM](https://github.com/sveltejs/sapper/issues/1093), on a per-route basis.
+1. [Faster cold-boot time for serverless functions](https://github.com/sveltejs/sapper/issues/356)
+1. [SPA Mode](https://github.com/sveltejs/sapper/issues/383)
+
+It would facilitate easier deployment on some new platforms:
+
+1. [Azure Functions](https://azure.microsoft.com/en-gb/services/functions)
+1. [AWS Lambda](https://aws.amazon.com/lambda/)
+1. [Google Cloud Functions](https://cloud.google.com/functions)
+1. [IBM Cloud Functions](https://www.ibm.com/uk-en/cloud/functions)
+1. [IPFS](https://ipfs.io/)
+1. [Begin.com](https://begin.com/)
+1. [Netlify](https://www.netlify.com/)
+1. [Cloudflare Workers](https://blog.cloudflare.com/introducing-cloudflare-workers/)
+1. [Fastly Edge](https://www.fastly.com/products/edge-compute)
+
+## Detailed design
+
+> This is the bulk of the RFC.
+
+> Explain the design in enough detail for somebody
+familiar with the framework to understand, and for somebody familiar with the
+implementation to implement. This should get into specifics and corner-cases,
+and include examples of how the feature is used. Any new terminology should be
+defined here.
+
+* Client-side bundle goes into static dir (SPA)
+* Individual modules built for each route
+* entrypoint imports individual route modules (for local development + simple `node` deployment)
+* `@sapper/[architecture]-builder` packages
+* Removing export (an application which has no dynamic routes is automatically SSG)
+
+Discuss:
+
+1. How do we determine which routes are:
+  * SPA (No preload)
+  * SSG (?)
+  * Hybrid / SSR (Preload function exists) - current behaviour
+
+2. Questions around treatment of server routes
+
+## How we teach this
+
+> What names and terminology work best for these concepts and why? How is this
+idea best presented? As a continuation of existing Svelte patterns, or as a
+wholly new one?
+
+> Would the acceptance of this proposal mean the Svelte guides must be
+re-organized or altered? Does it change how Svelte is taught to new users
+at any level?
+
+> How should this feature be introduced and taught to existing Svelte
+users?
+
+To define:
+
+SSG: Server-side generated pages
+SPA: Single-page App
+SSR: Server-side Rendering
+
+## Drawbacks
```

Comment:
One of the major questions this might raise is whether to keep support Webpack. Depending on the implementation, we may decide it's not worth implementing for Webpack (at least until someone from the community contributes a new Webpack implemention)

I think about 2/3 of our users use Rollup based on the numbers from the plugins:
* https://www.npmjs.com/package/rollup-plugin-svelte
* https://www.npmjs.com/package/svelte-loader

Reply:
I would ditch webpack entirely. We don't have the resource to support multiple bundlers, and 5 is probably different enough to require a re-implementation anyway. Rollup is fast becoming the preferred choice for other projects anyway, and we have a distinct advantage in having a certain project author on our team.

I am very much in favour of abstracting the build, I wouldn't build it for any specific bundler, but I also wouldn't go too deep in abstracting it.

Reply:
I am in favor of only focusing on rollup in the near-term, although I don't think we should allow any rollup-specific abstractions into the core algorithms of this stuff. This will: reduce the total amount of knowledge someone will need to understand the core algorithms, insulate us from changes to rollup, and make it easier to support other bundlers